compiler/rustc_hir_analysis/src/astconv/errors.rs

```diff
@@ -197,7 +201,34 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
         }
 
-        err.span_label(span, format!("associated type `{assoc_name}` not found"));
+        // If we still couldn't find any associated type, just list them all.
+
+        if all_candidate_names.is_empty() {
+            err.help(format!(
+                "`{ty_param_name}` has no associated type, try removing `{assoc_name}`"
+            ));
+            return err.emit();
+        }
+
+        let msg = if all_candidate_names.len() > 1 {
+            format!("`{ty_param_name}` has the following associated types")
+        } else {
+            format!("`{ty_param_name}` has the following associated type")
+        };
+
+        let applicability = if self.tcx().features().associated_type_defaults {
+            Applicability::Unspecified // `type A = Self::B` would suggest `type A = Self::A`
```

Comment:
What case is this protecting against?

Reply:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=848b04b36dea54f6539f37afd21fcd89
The compiler errors on `type A = Self::A`.